app/recordtransfer/validators.py

```diff
@@ -0,0 +1,159 @@
+    Categories used here:
+    - Uppercase (A-Z)
+    - Lowercase (a-z)
+    - Digits (0-9)
+    - Special characters from the allowed set: _ # % ( ) . ^ { } !.
+    """
+
+    def __init__(self, allowed_specials: str = "_#%().^{}!", required_categories: int = 3):
+        self.allowed_specials = allowed_specials
+        self.required_categories = required_categories
+
+    def validate(self, password: str, user: User | None = None) -> None:
+        """Validate that the password contains at least the required number of character
+        categories.
+        """
+        pwd = password or ""
+        has_upper = any(c.isupper() for c in pwd)
+        has_lower = any(c.islower() for c in pwd)
+        has_digit = any(c.isdigit() for c in pwd)
+        has_special = any(c in self.allowed_specials for c in pwd)
+        count = sum([has_upper, has_lower, has_digit, has_special])
+        if count < self.required_categories:
+            raise ValidationError(
+                _(
+                    "This password must contain at least three of the following: uppercase, lowercase, "
+                    "numbers, or one of these special characters: %(specials)s"
+                ),
+                code="password_not_enough_categories",
+                params={"specials": self.allowed_specials},
+            )
+
+    def get_help_text(self) -> str:
+        """Return help text describing the required character categories for the password."""
+        return _(
+            "Your password must include at least three of the following types: uppercase, lowercase, "
+            "numbers, or one of these special characters: _ # % ( ) . ^ { } !"
+        )
+
+
+class PasswordHistoryValidator:
+    """Ensure the new password is different from the user's previous N passwords."""
+
+    def __init__(self, history_depth: int = 5):
+        self.history_depth = history_depth
+
+    def validate(self, password: str, user: User | None = None) -> None:
+        """Validate that the password is not in the user's recent password history."""
+        if not user or not user.pk or not password:
+            return
+
+        # First check if the new password is the same as the current password
+        if user.password and check_password(password, user.password):
+            raise ValidationError(
+                _("Your new password cannot be the same as your current password."),
+                code="password_same_as_current",
+            )
+
+        # Then check against password history
+        recent_hashes = PasswordHistory.get_recent_password_hashes(
+            user=user, limit=self.history_depth
+        )
+        if any(check_password(password, old_hash) for old_hash in recent_hashes):
+            raise ValidationError(
+                _("You cannot reuse your previous %(n)d passwords."),
+                code="password_in_history",
+                params={"n": self.history_depth},
+            )
```

Comment:
It's conceivable that a system admin could set `self.history_depth=1`. Because of this, you should use `ngettext_lazy` instead.

```suggestion
            raise ValidationError(
                ngettext_lazy(
                    "You cannot reuse your previous password.",
                    "You cannot reuse your previous %(num)s passwords.",
                    self.history_depth,
                )
                % {
                    "num": self.history_depth,
                },
                code="password_in_history",
            )
```

I didn't recommend this for the password length validator because no sane system admin would set a minimum or maximum password length to only 1!